### PR TITLE
Mobile Manager Dynamic Menu Cell Updating

### PR DESCRIPTION
**Pull request by joeljfischer**
Update the naïve menu updating with a new dynamic add / removal system of menu cells to use a more robust method that intelligently adds and removes menu cells, thereby cutting down on the amount of RPCs needed to update an SDL menu.

### CODE REVIEW

**Comment by joeljfischer** [comment]
@jordynmackool This is ready for review

File: proposals/nnnn-mobile-dynamic-menu-cell-updating.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+## Proposed solution
+
+To solve this problem, we need a way to look at the old menu array, the new menu array, and determine adding / removing cells based on the differences.
+
+The first problem to be solved is to determine equality of cells. We don't currently do this for `SDLMenuCell`, but we do for `SDLChoiceCell`. The implementation of equality in `SDLMenuCell` will be very close to `SDLChoiceCell`, using the properties' values to create a hash and comparing those hashes.
+
+The second problem is to determine which adds and removals are necessary to turn the old / current menu array into the new menu array. We also need to consider sub-menus. Additionally complicating this issue is that existing menu cells cannot be moved, they must be removed and re-added, and if they have submenu cells, all those cells need to be removed and re-added as well.
```
Comment (jordynmackool) [inline proposals/nnnn-mobile-dynamic-menu-cell-updating.md]: Change to:

The second problem is to determine which adds and removals are necessary to turn the old / current menu array into the new menu array. Sub-menus will also need to be considered. Additionally, complicating this issue is the existing menu cells that cannot be moved. They must be removed and re-added, and if they have submenu cells, all those cells need to be removed and re-added as well.

File: proposals/nnnn-mobile-dynamic-menu-cell-updating.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+   
+   New Menu [B,A,D,C]
+   
+   B is marked for deletion and addition at position 0
+   
+   D is marked for deletion and addition at position 2
+
+#### Submenus
+
+Submenus work very similarly. Any menu cell which is staying in the menu which contains subcells will have the subcells checked for the same consistency as the top-level menu. Subcells will be marked for addition or deletion in the same way.
```
Comment (jordynmackool) [inline proposals/nnnn-mobile-dynamic-menu-cell-updating.md]: Change to:
Submenus work very similarly. Any menu cell which is staying in the menu that contains subcells will have the subcells checked for the same consistency as the top-level menu. Subcells will be marked for addition or deletion in the same way.

File: proposals/nnnn-mobile-dynamic-menu-cell-updating.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+If a menu cell is removed, all subcells will be removed by default because of the way that the `DeleteSubMenu` command works. Therefore, any cell that needs to be moved will need to have all of its submenus deleted.
+
+## Potential downsides
+
+1. This method is more intensive and complicated than the current method of deleting all old cells and adding all new cells. However, for simple changes (e.g. adding one menu item to the end of a menu of 80 commands), this method is vastly faster to send the singular `AddCommand` than to send 80 `DeleteCommand`s and 81 `AddCommand`s.
```
Comment (jordynmackool) [inline proposals/nnnn-mobile-dynamic-menu-cell-updating.md]: Change to:
1. This method is more intensive and complicated than the current method of deleting all old cells and adding all new cells. However, for simple changes (e.g. adding one menu item to the end of a menu of 80 commands) this method is vastly faster to send the singular `AddCommand` than to send 80 `DeleteCommand`s and 81 `AddCommand`s.

File: proposals/nnnn-mobile-dynamic-menu-cell-updating.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
+# Mobile Manager Dynamic Menu Cell Updating
+
+* Proposal: [SDL-NNNN](NNNN-mobile-dynamic-menu-cell-updating.md)
+* Author: [Joel Fischer](https://github.com/joeljfischer)
+* Status: **Awaiting review**
+* Impacted Platforms: [iOS / Android / Cloud]
+
+## Introduction
+
+Update the naïve menu updating with a new dynamic add / removal system of menu cells to use a more robust method that intelligently adds and removes menu cells, thereby cutting down on the amount of RPCs needed to update an SDL menu.
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply proposals/nnnn-mobile-dynamic-menu-cell-updating.md]: Leaving in "updating" is correct here. The menu is not what is naïve, it is how we update to a new menu that needs to change.

File: proposals/nnnn-mobile-dynamic-menu-cell-updating.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+## Motivation
+
+Currently, updating the menu through the menu manager in any way (even if the menu is only updated to add or remove a single cell) requires that the entire menu be deleted and added again. We should update our code to allow for updating the menu and dynamically determining which cells need to be deleted and which cells need to be added.
+
+## Proposed solution
+
+To solve this problem, we need a way to look at the old menu array, the new menu array, and determine adding / removing cells based on the differences.
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply proposals/nnnn-mobile-dynamic-menu-cell-updating.md]: Could you explain what you're looking for here? This is only the introductory sentence to the rest of the "proposed solution." The algorithm steps would provide the additional detail.

File: proposals/nnnn-mobile-dynamic-menu-cell-updating.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+* Status: **Awaiting review**
+* Impacted Platforms: [iOS / Android / Cloud]
+
+## Introduction
+
+Update the naïve menu updating with a new dynamic add / removal system of menu cells to use a more robust method that intelligently adds and removes menu cells, thereby cutting down on the amount of RPCs needed to update an SDL menu.
+
+## Motivation
+
+Currently, updating the menu through the menu manager in any way (even if the menu is only updated to add or remove a single cell) requires that the entire menu be deleted and added again. We should update our code to allow for updating the menu and dynamically determining which cells need to be deleted and which cells need to be added.
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply proposals/nnnn-mobile-dynamic-menu-cell-updating.md]: Fixed

File: proposals/nnnn-mobile-dynamic-menu-cell-updating.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+## Proposed solution
+
+To solve this problem, we need a way to look at the old menu array, the new menu array, and determine adding / removing cells based on the differences.
+
+The first problem to be solved is to determine equality of cells. We don't currently do this for `SDLMenuCell`, but we do for `SDLChoiceCell`. The implementation of equality in `SDLMenuCell` will be very close to `SDLChoiceCell`, using the properties' values to create a hash and comparing those hashes.
```
[the remark this answers is not in the document]
Comment (joeljfischer) [reply proposals/nnnn-mobile-dynamic-menu-cell-updating.md]: Fixed

File: proposals/nnnn-mobile-dynamic-menu-cell-updating.md (hunk 1 above)
Comment (joeljfischer) [reply proposals/nnnn-mobile-dynamic-menu-cell-updating.md]: Fixed

File: proposals/nnnn-mobile-dynamic-menu-cell-updating.md (hunk 2 above)
Comment (joeljfischer) [reply proposals/nnnn-mobile-dynamic-menu-cell-updating.md]: Fixed

File: proposals/nnnn-mobile-dynamic-menu-cell-updating.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jordynmackool) [reply proposals/nnnn-mobile-dynamic-menu-cell-updating.md]: Rephrase without `we `

To solve this problem, a way will need to be determined to look at the old menu array, the new menu array, and determine adding / removing cells based on the differences.

File: proposals/nnnn-mobile-dynamic-menu-cell-updating.md (hunk 3 above)
Comment (joeljfischer) [reply proposals/nnnn-mobile-dynamic-menu-cell-updating.md]: Fixed

File: proposals/nnnn-mobile-dynamic-menu-cell-updating.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (joeljfischer) [reply proposals/nnnn-mobile-dynamic-menu-cell-updating.md]: Fixed

**Comment by joeljfischer** [comment]
@jordynmackool Thank you for the review; all comments have been addressed.

**Pull request merged**